### PR TITLE
Add Memory tests

### PR DESCRIPTION
**Pull request by just-sparta**
### Tickets:
 - *51074*

### CODE REVIEW

File: tests/memory_tests/test_runner/conftest.py [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+    )
+    db_args_parser.addoption(
+        '--manifest',
+        type=Path,
+        required=is_db_used,
+        help='path to build manifest to extract commit information'
+    )
+
+
+@pytest.fixture(scope="session")
```
Comment (just-sparta) [inline tests/memory_tests/test_runner/conftest.py:127]: need to add fixture with OMZ models preparation

File: tests/memory_tests/test_runner/conftest.py [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+            if not report.passed:
+                instance["db"]["status"] = "failed"
+                instance["db"]["error_msg"] = report.longrepr.reprcrash.message
+            else:
+                instance["db"]["status"] = "passed"
+        instance["db"]["results"] = instance["results"]
+        instance["db"]["raw_results"] = instance["raw_results"]
+        logging.info("Upload data to {}/{}.{}. Data: {}".format(db_url, db_name, db_collection, instance["db"]))
+        # TODO: upload to new DB (memcheck -> memory_tests)
+        # upload_data(data, db_url, db_name, db_collection)
```
Comment (just-sparta) [inline tests/memory_tests/test_runner/conftest.py:473]: uncomment

Comment (just-sparta) [reply tests/memory_tests/test_runner/conftest.py:557]: it will be performed in the next task with working with the database

File: tests/memory_tests/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+set (CMAKE_BUILD_TYPE "Release" CACHE STRING "Choose the build type")
+
+project(memory_tests)
+
+set(OpenVINO_SOURCE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/../../")
+
+# Search OpenVINO Inference Engine installed
+find_package(InferenceEngine)
+
+if(NOT InferenceEngine_FOUND)
```
Comment (ishariko) [inline tests/memory_tests/CMakeLists.txt:16]: please use [CMakeList.txt](https://github.com/openvinotoolkit/openvino/blob/master/tests/stress_tests/CMakeLists.txt) as example for update current file

File: tests/memory_tests/README.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+``` bash
+mkdir build && cd build
+cmake .. && make memory_tests
+```
+
+If you don't have OpenVINO™ installed you need to have the `build` folder, which
+is created when you configure and build OpenVINO™ from sources:
+
+``` bash
+cmake .. -DInferenceEngine_DIR=$(realpath ../../../build) && make memory_tests
```
Comment (ishariko) [inline tests/memory_tests/README.md:26]: update README after changed CMakeLists.txt

File: tests/memory_tests/include/memory_tests_helper/utils.h [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+* @brief Get extension from filename
+* @param filename - name of the file which extension should be extracted
+* @return string with extracted file extension
+*/
+std::string fileExt(const std::string& filename) {
+    auto pos = filename.rfind('.');
+    if (pos == std::string::npos) return "";
+    return filename.substr(pos + 1);
+}
+}
```
Comment (ishariko) [inline tests/memory_tests/include/memory_tests_helper/utils.h:20]: add empty line in the end file

File: tests/memory_tests/src/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# Copyright (C) 2018-2021 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+
+add_subdirectory(memory_tests)
+add_subdirectory(memory_tests_helper)
```
Comment (ishariko) [inline tests/memory_tests/src/CMakeLists.txt:6]: add empty line

File: tests/memory_tests/src/memory_tests_helper/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+#
+
+set (TARGET_NAME "memory_tests_helper")
+
+file (GLOB SRC *.cpp)
+add_library(${TARGET_NAME} STATIC ${SRC})
+target_include_directories(${TARGET_NAME} PUBLIC "${CMAKE_SOURCE_DIR}/include")
+
+if(EXISTS "${OpenVINO_SOURCE_DIR}/thirdparty/gflags")
+    function(add_gflags)
```
Comment (ishariko) [inline tests/memory_tests/src/memory_tests_helper/CMakeLists.txt:12]: please rewrite getting gflags as example use [CMakeLists.txt](https://github.com/openvinotoolkit/openvino/blob/master/tests/stress_tests/common/CMakeLists.txt)

File: tests/memory_tests/src/memory_tests_helper/main.cpp [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+ */
+int main(int argc, char **argv) {
+  if (!parseAndCheckCommandLine(argc, argv))
+    return -1;
+
+  auto status =  _runPipeline();
+  StatisticsWriter::Instance().setFile(FLAGS_s);
+  StatisticsWriter::Instance().write();
+  return status;
+}
```
Comment (ishariko) [inline tests/memory_tests/src/memory_tests_helper/main.cpp:58]: add empty line

File: tests/memory_tests/src/memory_tests_helper/memory_counter.cpp [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+#include <memory>
+#include <string>
+#include <cstring>
+#include <vector>
+
+#include "statistics_writer.h"
+
+namespace MemoryTest {
+#ifdef _WIN32
+size_t getVmSizeInKB() {
```
Comment (ishariko) [inline tests/memory_tests/src/memory_tests_helper/memory_counter.cpp:16]: Update current file to the latest version of file [stress_tests/common/utils.cpp](https://github.com/openvinotoolkit/openvino/blob/master/tests/stress_tests/common/utils.cpp)

File: tests/memory_tests/src/memory_tests_helper/memory_counter.cpp [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+
+MemoryCounter::MemoryCounter(const std::string &mem_counter_name) {
+  name = mem_counter_name;
+  std::vector<size_t> memory_measurements = {getVmRSSInKB(), getVmHWMInKB(), getVmSizeInKB(),
+                                             getVmPeakInKB(), getThreadsNum()};
+  StatisticsWriter::Instance().addMemCounterToStructure({name, memory_measurements});
+}
+
+} // namespace MemoryTest
```
Comment (ishariko) [inline tests/memory_tests/src/memory_tests_helper/memory_counter.cpp:69]: empty line

**Review by ishariko — CHANGES_REQUESTED** [review]

File: tests/memory_tests/scripts/requirements.txt [hunk 11]
```diff
@@ -0,0 +1 @@
+PyYAML==5.4.1
```
Comment (ishariko) [inline tests/memory_tests/scripts/requirements.txt:1]: Maybe move this file to root dir? And use it in other requirement files?

File: tests/memory_tests/README.md [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+``` bash
+mkdir build && cd build
+cmake .. && make memory_tests
+```
+
+If you don't have OpenVINO™ installed you need to have the `build` folder, which
+is created when you configure and build OpenVINO™ from sources:
+
+``` bash
+cmake .. -DInferenceEngineDeveloperPackage_DIR=$(realpath ../../../build) && make memory_tests
```
Comment (ishariko) [inline tests/memory_tests/README.md:26]: and update `-DInferenceEngineDeveloperPackage` to `-DInferenceEngine`

**Review by alexander-shchepetov — APPROVED** [review]

**Review by omartiny — APPROVED** [review]

File: tests/memory_tests/src/memory_tests_helper/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+file (GLOB SRC *.cpp)
+add_library(${TARGET_NAME} STATIC ${SRC})
+target_include_directories(${TARGET_NAME} PUBLIC "${CMAKE_SOURCE_DIR}/include")
+
+add_subdirectory(${OpenVINO_SOURCE_DIR}/thirdparty/gflags
+                 ${CMAKE_CURRENT_BINARY_DIR}/gflags_build
+                 EXCLUDE_FROM_ALL)
+
+target_link_libraries(${TARGET_NAME}
```
Comment (ishariko) [inline tests/memory_tests/src/memory_tests_helper/CMakeLists.txt:15]: please create 1 line instead 3

**Review by ishariko — APPROVED** [review]

**Review by akoryach — APPROVED** [review]

File: tests/memory_tests/test_runner/conftest.py [hunk 14]
```diff
@@ -0,0 +1,564 @@
...
+@pytest.fixture(scope="function")
+def omz_models_conversion(pytestconfig, request):
+    """
+    Fixture for preparing omz models and updating test config with new paths
+    """
+    omz_repo = request.config.getoption("omz_repo")
+    if omz_repo:
+        omz_path = Path(omz_repo).resolve()
+    else:
+        omz_path = clone_omz_repo()
```
Comment (vurusovs) [inline tests/memory_tests/test_runner/conftest.py:205]: DON'T CLONE OMZ - use provided via CLI

**Review by vurusovs — CHANGES_REQUESTED** [review]

**Review by ishariko — APPROVED** [review]

File: tests/memory_tests/README.md (hunk 4 above)
Comment (vurusovs) [reply tests/memory_tests/README.md:26]: There should be no `InferenceEngine_DIR` option - need to set up OpenVINO correctly (install and call `setupvars` script) and after that build the tests

Refer to https://github.com/openvinotoolkit/openvino/blob/f5176f1c84b43fa89e29617e573f418983c63856/tests/time_tests/README.md

File: tests/memory_tests/README.md [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+If you don't have OpenVINO™ installed you need to have the `build` folder, which
+is created when you configure and build OpenVINO™ from sources:
+
+``` bash
+cmake .. -DInferenceEngine_DIR=$(realpath ../../../build) && make memory_tests
+```
+
+2. Run test:
+``` bash
+./scripts/run_memorytest.py ../../bin/intel64/Release/memtest_infer -m model.xml -d CPU
```
Comment (vurusovs) [inline tests/memory_tests/README.md:31]: executable won't be placed in `bin/` folder, so to determine a correct place need to `make install tests`

File: tests/memory_tests/include/memory_tests_helper/memory_counter.h [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <string>
+
+namespace MemoryTest {
+
+/** Encapsulate memory measurements.
```
Comment (vurusovs) [inline tests/memory_tests/include/memory_tests_helper/memory_counter.h:11]: docstring should be updated with new flow

File: tests/memory_tests/include/memory_tests_helper/utils.h [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+
+#include <string>
+
+namespace MemoryTest {
+/**
+* @brief Get extension from filename
+* @param filename - name of the file which extension should be extracted
+* @return string with extracted file extension
+*/
+std::string fileExt(const std::string& filename) {
```
Comment (vurusovs) [inline tests/memory_tests/include/memory_tests_helper/utils.h:15]: can be moved to global utils - may be done later

File: tests/memory_tests/src/memory_tests/CMakeLists.txt [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+add_subdirectory("${OpenVINO_SOURCE_DIR}/tests/lib" tests_shared_lib)
+
+foreach(test_source ${tests})
+    get_filename_component(test_name ${test_source} NAME_WE)
+    add_executable(${test_name} ${test_source})
+
+    target_link_libraries(${test_name} PRIVATE IE::inference_engine memory_tests_helper tests_shared_lib)
+
+    add_dependencies(memory_tests ${test_name})
```
Comment (vurusovs) [inline tests/memory_tests/src/memory_tests/CMakeLists.txt:20]: `install` is missed - https://github.com/openvinotoolkit/openvino/blob/master/tests/time_tests/src/timetests/CMakeLists.txt#L22

File: tests/memory_tests/test_runner/conftest.py [hunk 19]
```diff
@@ -0,0 +1,547 @@
...
+from platform_utils import get_os_name, get_os_version, get_cpu_info
+from utils import metadata_from_manifest, DATABASES, DB_COLLECTIONS
+
+MEMORY_TESTS_DIR = os.path.dirname(os.path.dirname(__file__))
+sys.path.append(MEMORY_TESTS_DIR)
+
+from test_runner.utils import query_memory_timeline, REFS_FACTOR
+
+
+OMZ_NUM_ATTEMPTS = 1
```
Comment (vurusovs) [inline tests/memory_tests/test_runner/conftest.py:47]: it should be greater to decrease number of problems related to network connection - refer to `get_testdata.py`

File: tests/memory_tests/test_runner/conftest.py [hunk 20]
```diff
@@ -0,0 +1,547 @@
...
+        "--niter",
+        type=check_positive_int,
+        help="number of iterations to run executable and aggregate results",
+        default=3
+    )
+    omz_args_parser = parser.getgroup("test with omz models")
+    omz_args_parser.addoption(
+        "--omz_repo",
+        type=Path,
+        required=True,
```
Comment (vurusovs) [inline tests/memory_tests/test_runner/conftest.py:86]: not required - models can be specified directly in config

File: tests/memory_tests/test_runner/conftest.py [hunk 21]
```diff
@@ -0,0 +1,547 @@
...
+    db_args_parser.addoption(
+        '--db_url',
+        type=str,
+        required=is_db_used,
+        help='MongoDB URL in a form "mongodb://server:port"'
+    )
+    db_args_parser.addoption(
+        '--timeline_report',
+        type=str,
+        required=is_db_used,
```
Comment (vurusovs) [inline tests/memory_tests/test_runner/conftest.py:138]: timeline uses DB, but DB not depend from timeline - so required=False

Comment (vurusovs) [reply tests/memory_tests/test_runner/conftest.py:138]: need to reorder options by it's meaning (e.g. --timeline_report and --mo_path)

File: tests/memory_tests/test_runner/conftest.py [hunk 22]
```diff
@@ -0,0 +1,547 @@
...
+        help="Path to Open Model Zoo (OMZ) repository.",
+    )
+    omz_args_parser.addoption(
+        "--omz_models_out_dir",
+        type=Path,
+        default=abs_path('../_omz_out/models'),
+        help="Directory to put test data into.",
+    )
+    omz_args_parser.addoption(
+        "--mo_path",
```
Comment (vurusovs) [inline tests/memory_tests/test_runner/conftest.py:96]: just --mo

File: tests/memory_tests/test_runner/conftest.py [hunk 23]
```diff
@@ -0,0 +1,547 @@
...
+    )
+    test_args_parser.addoption(
+        "--niter",
+        type=check_positive_int,
+        help="number of iterations to run executable and aggregate results",
+        default=3
+    )
+    omz_args_parser = parser.getgroup("test with omz models")
+    omz_args_parser.addoption(
+        "--omz_repo",
```
Comment (vurusovs) [inline tests/memory_tests/test_runner/conftest.py:84]: just --omz

File: tests/memory_tests/test_runner/conftest.py [hunk 24]
```diff
@@ -0,0 +1,547 @@
...
+        type=Path,
+        required=True,
+        help="Path to Open Model Zoo (OMZ) repository.",
```
Comment (vurusovs) [inline tests/memory_tests/test_runner/conftest.py:87]: to repository root

File: tests/memory_tests/test_runner/conftest.py [hunk 25]
```diff
@@ -0,0 +1,547 @@
...
+    cache_dir = request.config.getoption("omz_cache_dir")
+    omz_models_out_dir = request.config.getoption("omz_models_out_dir")
+    omz_irs_out_dir = request.config.getoption("omz_irs_out_dir")
+    mo_path = request.config.getoption("mo_path")
+
+    downloader_path = omz_path / "tools" / "downloader" / "downloader.py"
+    converter_path = omz_path / "tools" / "downloader" / "converter.py"
+    info_dumper_path = omz_path / "tools" / "downloader" / "info_dumper.py"
+
+    records = [rec for rec in pytestconfig.session_info]
```
Comment (vurusovs) [inline tests/memory_tests/test_runner/conftest.py:199]: since fixture has scope=function, fixture status will be a part of test status. In this case fixture should prepare only the model from specific test, not for all run

I suggest doing conversion model per test in order to spread conversion status to the test status - it will help to maintain issues related to OMZ or MO easier

File: tests/memory_tests/test_runner/conftest.py [hunk 26]
```diff
@@ -0,0 +1,547 @@
...
+    with timetests results saved as references.
+    """
+    yield
+    new_tconf_path = pytestconfig.getoption('dump_refs')
+    if new_tconf_path:
+        logging.info("Save new test config with test results as references to {}".format(new_tconf_path))
+
+        upd_cases = []
+        steps_to_dump = {"create_exenetwork", "first_inference"}
+        vm_metrics_to_dump = {"vmhwm", "vmpeak", "vmrss", "vmsize"}
```
Comment (vurusovs) [inline tests/memory_tests/test_runner/conftest.py:456]: VMRSS and VMHWM only - other isn't required for now

File: tests/memory_tests/test_runner/html.html [hunk 27]
```diff
@@ -0,0 +1,1606 @@
+<!DOCTYPE html>
```
Comment (vurusovs) [inline tests/memory_tests/test_runner/html.html:1]: to be removed

File: tests/memory_tests/test_runner/requirements.txt [hunk 28]
```diff
@@ -0,0 +1,8 @@
+pytest~=5.0
+attrs==19.1.0
```
Comment (vurusovs) [inline tests/memory_tests/test_runner/requirements.txt:2]: `attrs` was required for other `pytest` version to provide compatibility - shouldn't required for now and can be removed

File: tests/memory_tests/test_runner/test_config.yml [hunk 29]
```diff
@@ -0,0 +1,14 @@
+- device:
+    name: CPU
+  model:
+    path: ${SHARE}/stress_tests/master_04d6f112132f92cab563ae7655747e0359687dc9/caffe/FP32/alexnet/alexnet.xml # TODO: add link to `test_data` repo model
```
Comment (vurusovs) [inline tests/memory_tests/test_runner/test_config.yml:4]: use timetests IRs, it's more recent

File: tests/memory_tests/test_runner/utils.py [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+    """Compare values with provided reference"""
+
+    vm_metrics_to_compare = {"vmrss", "vmhw"}
+    status = 0
+
+    for step_name, vm_records in reference.items():
+        for vm_metric, stat_metrics in vm_records.items():
+            if vm_metric not in vm_metrics_to_compare:
+                continue
+            for stat_metric_name, reference_val in stat_metrics.items():
```
Comment (vurusovs) [inline tests/memory_tests/test_runner/utils.py:93]: note that only `avg` will be dumped to references in current implementation. So it can lead to key error if iterates by current stats keys, or missed if iterates by references keys

File: tests/time_tests/test_runner/conftest.py [hunk 31]
```diff
@@ -78,11 +77,18 @@ def pytest_addoption(parser):
         required=is_db_used,
         help='MongoDB URL in a form "mongodb://server:port"'
     )
+    db_args_parser.addoption(
+        '--db_name',
```
Comment (vurusovs) [inline tests/time_tests/test_runner/conftest.py:81]: not required, can be hardcoded for timetests

File: tests/time_tests/scripts/run_timetest.py [hunk 32]
```diff
@@ -27,35 +26,11 @@
 
 from test_runner.utils import filter_timetest_result
 
+UTILS_DIR = os.path.join(Path(__file__).parent.parent.parent, "utils")
```
Comment (vurusovs) [inline tests/time_tests/scripts/run_timetest.py:29]: I would prefer to do timetests refactoring in separate PR in order to not to combine a lot of changes here

**Review by vurusovs — CHANGES_REQUESTED** [review]
If timetests changes will be moved to separate PR, approve to merge. But later a lot of comments should be fixed
